File: leetcode/#53.js
// question --> https://leetcode.com/problems/maximum-subarray/

var maxSubArray = function(nums) {
    let maxSoFar = nums[0];
    let maxEndingHere = nums[0];
    for (let index = 1; index < nums.length; index++) {
        let temp = maxEndingHere + nums[index];
        maxEndingHere = Math.max(temp, nums[index]);
        maxSoFar = Math.max(maxSoFar, maxEndingHere);
    }
    return maxSoFar
};

// TC: O(N); SC: O(1)

// Alternative Approach

/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    let summ = 0; 
    let ans = nums.reduce(function(a, b) {
        return Math.max(a, b);
    })
    if (ans < 0) {
        return ans; 
    } else {
        ans = 0; 
    }
    
    for (let i = 0; i < nums.length; i++) {
        summ += nums[i]; 
        if (summ < 0) {
            summ = 0; 
        }
        ans = Math.max(ans, summ) 
    }
    return ans; 
};

// end

File: leetcode/#53.py
# question --> https://leetcode.com/problems/maximum-subarray/

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        max_so_far, max_ending_here = nums[0], nums[0]
        
        for index in range(1, len(nums)):
            temp = max_ending_here + nums[index]
            max_ending_here = max(temp, nums[index])
            max_so_far = max(max_so_far, max_ending_here)
            
        return max_so_far
        
        # TC: O(N); SC: O(1) 
        
# Alternative Approach

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        summ = 0
        ans = max(nums)
        if ans < 0:
            return ans
        else:
            ans = 0

        for number in nums:
            summ += number
            if summ < 0: 
                summ = 0
            ans = max(ans, summ)
        
        return ans         
        
        # TC: O(N); SC: O(1)
        
# end 